### PR TITLE
Revpi 2487/pb serdev

### PR DESCRIPTION
**Pull request by linosanfilippo-kunbus**

### CODE REVIEW

File: drivers/tty/serdev/pibridge.c [hunk 1]
```diff
@@ -175,12 +176,14 @@ int pibridge_recv_timeout(u8 *buf, u16 len, u16 timeout)
...
 	return ret;
 }
+EXPORT_SYMBOL(pibridge_recv_timeout);
 
 int pibridge_recv(u8 *buf, u16 len)
 {
 	/* using default timeout REV_PI_IO_TIMEOUT */
 	return pibridge_recv_timeout(buf, len, REV_PI_IO_TIMEOUT);
 }
+EXPORT_SYMBOL(pibridge_recv);
```
Comment (zhan-kunbus) [inline drivers/tty/serdev/pibridge.c:185]: it maybe not needed for pibridge_recv to export, as pibridge_recv_timeout should be enough

**Review by l1k — COMMENTED** [review]
Commit 159bf378d363fd39052067735e29bde07c2f6f76 contains an unrelated change in `net/dsa/slave.c`, looks like this seeped in from the other pull request...

**Comment by l1k** [comment]
Not sure if it makes sense to define the `current-speed` in the DT. The documentation in `Documentation/devicetree/bindings/serial/slave-device.txt` says:

<blockquote>
                  This should
                  only be present in case a driver has no chance to know
                  the baud rate of the slave device.
                  Examples:
                    * device supports auto-baud
                    * the rate is setup by a bootloader and there is no
                      way to reset the device
                    * device baud rate is configured by its firmware but
                      there is no way to request the actual settings
</blockquote>

Normally the DT only describes things baked into the hardware. In this case, we know the baudrate is 115200 and we could set that directly in the driver. The driver could in theory negotiate a higher baudrate with devices if we'd amend their firmware to support it.

**Comment by l1k** [comment]
Still in commit 159bf37, the `default n` Kconfig entry can be omitted as all options are disabled by default unless they're explicitly enabled (or selected by another config option).

**Comment by linosanfilippo-kunbus** [comment]
> Commit [159bf37](https://github.com/RevolutionPi/linux/commit/159bf378d363fd39052067735e29bde07c2f6f76) contains an unrelated change in `net/dsa/slave.c`, looks like this seeped in from the other pull request...

Oops, indeed. I will update this pull request.

**Comment by l1k** [comment]
```
+        depends on TTY
+        depends on SERIAL_DEV_BUS != m
```

I think we just copy-pasted this from `SERIAL_DEV_CTRL_TTYPORT` but it doesn't really make sense. What we actually depend on is `SERIAL_DEV_BUS` (just like in `drivers/gnss/Kconfig`), but since this is already enclosed in an `if SERIAL_DEV_BUS`, we don't need any further dependencies. So the two above can be dropped.

**Comment by l1k** [comment]
`#include` directives should be sorted alphabetically.

**Comment by linosanfilippo-kunbus** [comment]
Ok, I will adjust the files and push again, thx!

**Comment by l1k** [comment]
DT vendor id `kunbus` needs to be added to `Documentation/devicetree/bindings/vendor-prefixes.yaml`.

**Comment by linosanfilippo-kunbus** [comment]
> `#include` directives should be sorted alphabetically.

I think there is still a lot of polishing needed later when we actually want to submit this to mainline. Lets do this in a separate task. For now I suggest to only focus on functional correctness.

**Comment by l1k** [comment]
Okay I've verified that the termios settings in `piIoComm_open_serial()` are identical to those in `ttyport_open()` + `pibridge_parse_dt()`. Nice. Only exception is `CREAD` which is set by the former but not the latter. Should be harmless though.

File: drivers/tty/serdev/pibridge.c [hunk 2]
```diff
@@ -57,7 +57,6 @@ static int pibridge_parse_dt(struct serdev_device *serdev)
 
 	of_property_read_u32(node, "current-speed", &speed);
 	serdev_device_set_baudrate(serdev, speed);
-
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c]: Actually I think this newline makes sense to separate the operations performed on the serial port, so I'd prefer if it could be kept.

File: drivers/tty/serdev/pibridge.c [hunk 3]
```diff
@@ -253,64 +250,64 @@ int pibridge_req_gate_tmt(u8 dst, u16 cmd, u8 *snd_buf, u16 snd_len,
...
 	/* Do not wait for a response in case of a broadcast address */
 	if (dst == 0xff)
 		return 0;
 
-	if (sizeof(pkthdr) != pibridge_recv_timeout((u8 *)&pkthdr, sizeof(pkthdr), tmt)){
+	if (sizeof(pkthdr) != pibridge_recv_timeout((u8 *) &pkthdr,
+						    sizeof(pkthdr), tmt)) {
 		pb_err(25, "receive head error in gate-req(hdr_len:%d, timeout:%d, data0:%c)\n",
-				sizeof(pkthdr), tmt, snd_buf?snd_buf[0]:0);
+		       sizeof(pkthdr), tmt, snd_buf?snd_buf[0]:0);
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:264]: Add blanks to the ternary operator.

File: drivers/tty/serdev/pibridge.c [hunk 4]
```diff
@@ -375,16 +373,17 @@ int pibridge_req_io(u8 addr, u8 cmd, u8 *snd_buf, u16 snd_len, u8 *rcv_buf, u16
...
 	if (pibridge_req_send_io(addr, cmd, snd_buf, snd_len)) {
 		pb_err(25, "send message error in io-req(addr:%d, cmd:%d, len:%d)\n",
-				addr, cmd, snd_len);
+		       addr, cmd, snd_len);
 		return -EIO;
 	}
 
-	if (sizeof(pkthdr) != pibridge_recv((u8 *)&pkthdr, sizeof(pkthdr))) {
+	if (sizeof(pkthdr) !=
+	    pibridge_recv((u8 *) &pkthdr, sizeof(pkthdr))) {
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:380]: I'd swap the arguments, i.e. call `pibridge_recv()` and compare the result to the expected value `sizeof(pkthdr)`. I think inverting the arguments as is done here is frowned upon in the kernel community. This was likely inadvertently carried over from piControl code.

File: drivers/tty/serdev/pibridge.c [hunk 5]
```diff
@@ -184,14 +184,13 @@ int pibridge_recv(u8 *buf, u16 len)
...
 
-#pragma pack(1)
 struct pibridge_pkthdr_gate {
 	u8	dst;
 	u8	src;
 	u16	cmd;
 	u16	seq;
 	u8	len;
-};
+} __attribute__((packed));
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:193]: Declare `struct pibridge_pkthdr_io` packed as well while you're at it.

File: drivers/tty/serdev/pibridge.c [hunk 6]
```diff
@@ -116,15 +116,15 @@ static void pibridge_remove(struct serdev_device *serdev)
 
 #ifdef CONFIG_OF
 static const struct of_device_id pibridge_of_match[] = {
-	{ .compatible = "kunbus,pibridge" },
+	{ .compatible = "kunbus,pibridge-comm" },
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:117]: I think this is wrong. Handling of the sniff pins will have to be moved to this driver so that it can perform the bus scan and instantiate devices on the PiBridge bus. This is really "the" PiBridge driver. If piControl toggles the sniff pins for now, rename the node to which the piControl platform device driver binds.

The only other thing that's done on the PiBridge is the Gateway communication but that's completely separate.

File: drivers/tty/serdev/pibridge.c [hunk 7]
```diff
@@ -259,6 +259,9 @@ int pibridge_req_gate_tmt(u8 dst, u16 cmd, u8 *snd_buf, u16 snd_len,
 				dst, cmd, snd_len);
 		return -EIO;
 	}
+	/* Do not wait for a response in case of a broadcast address */
+	if (dst == 0xff)
+		return 0;
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:259]: Interesting. Does the existing code in piControl do the same?

File: drivers/tty/serdev/pibridge.c [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+
+#define REV_PI_IO_TIMEOUT           10         // msec
+#define REV_PI_RECV_BUFFER_SIZE     100
+
+#define pb_err(n, fmt, ...) \
+	do {\
+		static int cc = 0;\
+		if (cc < n && pibridge_s ){\
+			dev_err(&pibridge_s->serdev->dev, fmt, ##__VA_ARGS__);\
+			cc ++;\
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:17]: Hm, this is called with "n = 25" everywhere.

@zhan-kunbus What's the motivation for limiting these error messages to 25 and dropping all of them once that limit is reached?

Maybe `dev_err_ratelimited()` would be more appropriate?

File: drivers/tty/serdev/pibridge.c [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+#pragma pack(1)
+struct pibridge_pkthdr_gate{
+	u8	dst;
+	u8	src;
+	u16	cmd;
+	u16	seq;
+	u8	len;
+};
+
+u8 pibridge_crc8(u8 base, u8 *data, u16 len)
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:195]: Can be declared `static`. Probably applies to other functions as well.

File: drivers/tty/serdev/pibridge.c [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+
+	return 0;
+}
+EXPORT_SYMBOL(pibridge_req_send_gate);
+
+#define PIBRIDGE_RESP_CMD	0x3fff
+#define PIBRIDGE_RESP_OK	0x4000
+#define PIBRIDGE_RESP_ERR	0x8000
+
+int pibridge_req_gate_tmt(u8 dst, u16 cmd, u8 *snd_buf, u16 snd_len,
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:245]: For now, passing a destination address to these functions and having a global `pibridge_s` instance is fine.

Once we perform a bus scan and instantiate a `struct pibridge_dev` for each module found, we'll pass that struct to the send/receive functions and it'll contain both the module address and the pointer to the `struct pibridge`.

File: drivers/tty/serdev/pibridge.c [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+	.driver	= {
+		.name		= "pibridge",
+		.of_match_table	= of_match_ptr(pibridge_of_match),
+	},
+	.probe	= pibridge_probe,
+	.remove	= pibridge_remove,
+};
+module_serdev_device_driver(pibridge_driver);
+
+
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:133]: Superfluous newline.

File: drivers/tty/serdev/pibridge.c [hunk 12]
```diff
@@ -0,0 +1,409 @@
...
+	int ret;
+
+	ret = serdev_device_write(serdev, buf, len, MAX_SCHEDULE_TIMEOUT);
+	if (ret < 0)
+		pb_err(25, "write to serdev error(len:%d)\n", len);
+
+	/* read fifo may contain stale bytes */
+	mutex_lock(&pi->lock);
+	kfifo_reset(&pi->read_fifo);
+	mutex_unlock(&pi->lock);
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:149]: Clearing the read FIFO *after* sending data looks dangerous because it assumes that it finishes before the module sends an answer. It also feels wrong if `pibridge_send()` clears the READ FIFO as a side effect. I think this should happen explicitly in those functions that receive data, e.g. `pibridge_req_gate_tmt()`, either at the top of such functions or before sending out the last byte.

File: drivers/tty/serdev/pibridge.c [hunk 13]
```diff
@@ -0,0 +1,409 @@
...
+int pibridge_recv_timeout(u8 *buf, u16 len, u16 timeout)
+{
+	struct pibridge *pi = pibridge_s;
+	int jiffies;
+	int ret;
+
+	jiffies = wait_event_timeout(pi->read_queue, kfifo_len(&pi->read_fifo) >= len,
+			msecs_to_jiffies(timeout));
+	mutex_lock(&pi->lock);
+	if (jiffies){
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c]: The if-condition looks wrong: It would seem more logical to re-check for `kfifo_len(&pi->read_fifo) >= len` because the desired number of bytes may have been received after the timeout expired and before the lock was acquired.

File: drivers/tty/serdev/pibridge.c [hunk 14]
```diff
@@ -0,0 +1,409 @@
...
+#define PIBRIDGE_RESP_OK	0x4000
+#define PIBRIDGE_RESP_ERR	0x8000
+
+int pibridge_req_gate_tmt(u8 dst, u16 cmd, u8 *snd_buf, u16 snd_len,
+		u8 *rcv_buf, u16 rcv_len, u16 tmt)
+{
+	struct pibridge_pkthdr_gate pkthdr;
+	u8 crc_rcv;
+	u8 crc;
+
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:251]: Superfluous newline.

File: drivers/tty/serdev/pibridge.c [hunk 15]
```diff
@@ -0,0 +1,409 @@
...
+	if (rcv_len != 0){
+		crc = pibridge_crc8(crc, rcv_buf, rcv_len);
+	}
+	if (crc != crc_rcv){
+		return -EBADMSG;
+	}
+
+	if (!(pkthdr.cmd & PIBRIDGE_RESP_OK)){
+		pb_err(25, "bad responsed OK code in gate-req(cmd:%d)\n",
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:290]: "responsed" is not proper English.

File: drivers/tty/serdev/pibridge.c [hunk 16]
```diff
@@ -0,0 +1,409 @@
...
+		u8 *rcv_buf, u16 rcv_len)
+{
+	return pibridge_req_gate_tmt(dst, cmd, snd_buf, snd_len,
+			rcv_buf, rcv_len, REV_PI_IO_TIMEOUT);
+}
+EXPORT_SYMBOL(pibridge_req_gate);
+struct pibridge_pkthdr_io
+{
+	u8 addr	:6;
+	u8 typ	:1;	/* 0 for unicast, 1 for broadcast */
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c:325]: type

File: drivers/tty/serdev/pibridge.c (hunk 2 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c]: Ok

File: drivers/tty/serdev/pibridge.c (hunk 3 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:264]: Ok

File: drivers/tty/serdev/pibridge.c (hunk 4 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:380]: Agreed

File: drivers/tty/serdev/pibridge.c (hunk 5 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:193]: Indeed, I must have overlooked this, thx.

File: drivers/tty/serdev/pibridge.c (hunk 6 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:117]: Huh? IIRC our plan was to implement a pibridge bus driver. Device detection is definitely a job for a bus driver. The bus driver then uses the serdev for communication but nothing else...

File: drivers/tty/serdev/pibridge.c (hunk 7 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:259]: Yes. Would be better to use a constant/makro for the "0xff", though

File: drivers/tty/serdev/pibridge.c (hunk 8 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:17]: AFAIK those custom functions to print messages are not well received by some maintainers

File: drivers/tty/serdev/pibridge.c (hunk 9 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:195]: Agreed

File: drivers/tty/serdev/pibridge.c (hunk 1 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:185]: For now I only made use of the low-level functions `pibridge_send()` and `pibridge_recv_timeout()` (the only exception is `pibridge_req_gate_tmt()` since I found it easier to replace the current code in this case). However this was only to have as minimal impact (i.e. changes) to the current code as possible. I do not think that the high-level functions are worthless though, maybe we can use it later when we implement the new piControl design.

File: drivers/tty/serdev/pibridge.c (hunk 12 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:149]: Yes, the FIFO should be cleared before sending not after. I also agree that `pibridge_send()` is the wrong place to do so. However I dont think that doing this in one of the receive functions is correct either. IMHO it should be done immediately before `pibridge_send()` is called. For this reason there should be an own function like `clear_buffer()`. That function can be called in higher-level functions like `pibridge_req_send_gate()` or `pibridge_req_send_io()` before `pibridge_send()` is called.

**Comment by l1k** [comment]
<blockquote>
For now I only made use of the low-level functions pibridge_send() and pibridge_recv_timeout() (the only exception is pibridge_req_gate_tmt() since I found it easier to replace the current code in this case). However this was only to have as minimal impact (i.e. changes) to the current code as possible. I do not think that the high-level functions are worthless though, maybe we can use it later when we implement the new piControl design.
</blockquote>

piDIOComm.c / piAIOComm.c construct IO protocol packets manually, then send/receive them via `piIoComm_send()` / `piIoComm_recv()`. Those other functions you're not using yet are meant to be used there.

**Comment by linosanfilippo-kunbus** [comment]
> > For now I only made use of the low-level functions pibridge_send() and pibridge_recv_timeout() (the only exception is pibridge_req_gate_tmt() since I found it easier to replace the current code in this case). However this was only to have as minimal impact (i.e. changes) to the current code as possible. I do not think that the high-level functions are worthless though, maybe we can use it later when we implement the new piControl design.
> 
> piDIOComm.c / piAIOComm.c construct IO protocol packets manually, then send/receive them via `piIoComm_send()` / `piIoComm_recv()`. Those other functions you're not using yet are meant to be used there.

Yes right. But using the high level functions meant to adjust the piControl code at much more places, always with the danger to introduce side effects. My approach was to only make minimal changes by replacing only the low level functions. Later when we reimplement the caller of the serdev (be it piControl or whatever the module may be called) we can use the high level functions of course, since the send-receive logic does not have to be implemented in the caller. This is the reason why I kept these functions exported although they are not used right now.

File: drivers/tty/serdev/pibridge.c (hunk 8 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:17]: > AFAIK those custom functions to print messages are not well received by some maintainers

Also I am  sure that accessing variables in a macro that have to be defined by the caller (like pibridge_s in pb_err()) is frawned upon. I vote for not printing any messages at all in the serdev module. Instead lets define some sane return values for error conditions and then let the caller decide how to handle them.

File: drivers/tty/serdev/pibridge.c (hunk 13 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c]: I would not call the current code wrong, since we specified a timeout and therefore can treat the expiration as an error. But yes, your suggestion would be a slight improvement of the current implementation (also in the current implementation we do not have to take the mutex before the "if" condidtion).

**Comment by l1k** [comment]
<blockquote>Yes right. But using the high level functions meant to adjust the piControl code at much more places, always with the danger to introduce side effects. My approach was to only make minimal changes by replacing only the low level functions. Later when we reimplement the caller of the serdev (be it piControl or whatever the module may be called) we can use the high level functions of course, since the send-receive logic does not have to be implemented in the caller. This is the reason why I kept these functions exported although they are not used right now.</blockquote>

Right, calling the high level functions can be postponed and need not be done as part of this pull. However it would be the next logical step in the endeavour to refactor piControl.

File: drivers/tty/serdev/pibridge.c (hunk 6 above)
Comment (l1k) [reply drivers/tty/serdev/pibridge.c:117]: There seems to be a misunderstanding. This *is* the pibridge bus driver, or the first portion of it.

For now, this driver only provides a set of functions to perform RS-485 communication. There's still some work to do after this pull request is merged to convert all functions in piControl over to the new RS-485 communication primitives.

The next step will be to add bus scan functionality to the driver (using the sniff pins). We'll then define a `struct pibridge_dev` (which embeds a `struct device`) and instantiate one for each module found. That will allow us to change the communication primitives such that they're no longer invoked with a bus address parameter (an integer), but with a `struct pibridge_dev` parameter. The struct contains the bus address among other things.

As a third step, we can define individual device types for DIO, AIO, MIO and assign that type to the `struct device` which is instantiated for each found PiBridge device. That will allow us to write drivers for each module which bind to the device of same type.

To summarize, what this driver will be doing is:
* scan the bus using the sniff pins, instantiate devices
* provide RS-485 communication primitives which are used by the devices' drivers

File: drivers/tty/serdev/pibridge.c (hunk 12 above)
Comment (l1k) [reply drivers/tty/serdev/pibridge.c:149]: Yes this looks much saner now and is also more efficient. Before, the receive FIFO was cleared multiple times.

I think the reason the FIFO was cleared after sending was to suppress the own echo. Nowadays we suppress the echo in `amba-pl011.c` by disabling the receiver upon asserting RTS. But it wasn't always like this. RS-485 support wasn't added to `amba-pl011.c` until late 2017. The Core was already on the market for a year at that point. The echo was supposed to be suppressed by the hardware circuit for Transmit Enable assertion but I think that didn't work perfectly. Hence the need to clear the echo in piControl.

File: drivers/tty/serdev/pibridge.c (hunk 6 above)
Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:117]: AFAIU for a bus driver we have to define a `bus_type` and then register it with `bus_register()`. Will this be done in this file (pibridge.c) then?

Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:117]: I think my understanding was: We have a kernel module which defines a `bus_type` "pibridge" (or something like that), which we register with `bus_register()`. Then in the `probe()` function this module discovers all physically attached devices by using the sniff pins. For each found device as struct `device` is instantiated and registered with `device_register()` (this fills the `klist_devices` member of the struct `subsys_private` member of `bus_type`). 
As soon as a driver for one (or more if of the same type) of the devices is registered `bus_type->match()` is called and checks the registered driver against the list of found devices (a match fills the list of `klist_drivers` in the `subsys_private` struct). If a match is found the `probe()` function of the device driver for the specific device is called (done by the device core). In that function the device driver sends its configuration data via RS485. For this it uses the library functions provided in pibridge.c (which for that reason better should be called pibridge-comm.c and which is the reason why I chose "kunbus,pibridge-comm" as the compatible name). So in my understanding the pibride communiction module (aka pibridge.c) and the bus driver module are different modules...

Comment (l1k) [reply drivers/tty/serdev/pibridge.c:117]: <blockquote>
AFAIU for a bus driver we have to define a bus_type and then register it with bus_register(). Will this be done in this file (pibridge.c) then?
</blockquote>

Yes. It could be in a separate file if you think that it becomes too large but that other file and `pibridge.c` would be linked together to form a driver which performs the RS-485 communication and provides a bus on top of it. Including device enumeration on that bus using the sniff pins. It's all part of that single driver. So the sniff pins handling would also be moved here eventually.

In the device hierarchy, there's the UART device, plus the pibridge platform device below the UART, plus the bus below the pibridge platform device, plus all the module devices on that bus.

Comment (l1k) [reply drivers/tty/serdev/pibridge.c:117]: You can call the *file* `pibridge-comm.c`, but  not the compatible string. There are not two separate platform devices in the device hierarchy (one for the RS-485 communication and one for the bus on top of it). At least I couldn't see a benefit of such a separation.

Comment (linosanfilippo-kunbus) [reply drivers/tty/serdev/pibridge.c:117]: What about the approach I sketched above (with a pibridge-comm library module and a bus driver module)?

Comment (l1k) [reply drivers/tty/serdev/pibridge.c:117]: Only makes sense if you intend to let the bus driver modul run on top of some other communication protocol. (PiBridge 2.0?) Otherwise just adds overhead and confusion IMO.
